### ISSUE TITLE
Question about initialization of components in set_params()

### ISSUE THREAD

**Issue by gtauzin** [skorch-dev/skorch#784]
Hello,
First of all, thank you so much for this library. I find the sklearn ecosystem interface super valuable and I love how easy to use you made it!

I am curious about some of the technicalities behind the interfacing. The main one is that I noticed that the `set_params` function also take care of reinitializing any of the necessary component, while the `partial_fit` or `fit` method actually call the initialization of all components as well. In the sklearn API, I often read that all attributes should be set in `fit` and indeed in `GridSearchCV` sklearn will set the parameters and then systematically call `fit` right after. I imagine, this reinitialization happens in `set_params` for a reason. Would you mind explaining it?

Thank you for work~ :)

**Comment by BenjaminBossan** [comment]
Thanks, it's always nice to hear that people enjoy using skorch.

As to your question: `set_params` needs to initialize certain components because the parameter that is set might require it. E.g., let's say a parameter for the number of units in a linear layer is changed. Then we cannot re-use the same module instance, we have to initialize it again, with the weights having a new shape.

There are some arguments that don't actually require re-initialization. E.g. when you change the learning rate or the batch size, it's not necessary and it won't happen. However, if you change an argument on the module, skorch cannot know if the module needs to be re-initialized or not, therefore it always re-initializes.

Recently, there has been a minor change to this logic. When the net has not been initialized yet -- by which I mean it was instantiated but neither of `fit`, `partial_fit`, or `initialize` has been called -- and then you call `set_params`, the neot is no longer (re-)initialized. This change can be found [here](https://github.com/skorch-dev/skorch/blame/1165a786b6173fe85ad58899d6c81f13bb596164/skorch/net.py#L1920-L1925).

I hope this answers your question.

**Comment by gtauzin** [comment]
Hi @BenjaminBossan and thank you so much for taking the time to give me a detailed answer. It is very clear :)

Thank you again for your work :)

**Comment by ottonemo** [comment]
As an addendum: as @BenjaminBossan mentioned some parameters, like the learning rate, do not re-initialize the components they are referring to. This is done via "virtual" parameters and if you need such a parameter (e.g., to set a parameter on the module without re-initialization because you know what you are doing) you can use virtual parameters for that.

Note, though, that there is not yet a public API for virtual parameters as we are still figuring out whether this is something we want to support, so this API cannot be relied upon :)
